### PR TITLE
Add Hinge metric

### LINKED ISSUE #89: [New Metric] Hinge losses
## 🚀 Feature

Add Hinge losses to classification package:
https://scikit-learn.org/stable/modules/generated/sklearn.metrics.hinge_loss.html#sklearn.metrics.hinge_loss
possibly with an added parameter `squared` to also calculated squared hinge loss

### PR DESCRIPTION
**Pull request by ethanwharris**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  

## What does this PR do?
Fixes #89 - added the `multiclass_mode` parameter to enable computation in a one-vs-all fashion (e.g. for multi-class SVMs)

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by Borda — COMMENTED** [review]
I am not very sure about adding anything named as Loss, could we have it consistent with others?

**Comment by ethanwharris** [comment]
@Borda Happy to change if needed :smiley: although not sure what else to call it. I guess the options are:
- `hinge_loss` (consistent with sklearn)
- `hinge` (avoids "loss")
- it's sometimes also called the max-margin objective, so maybe that?

**Comment by Borda** [comment]
I would say `hinge` is fine...
cc: @PyTorchLightning/core-metrics other suggestions?

**Comment by SkafteNicki** [comment]
+1 for `hinge` :]

File: docs/source/references/functional.rst [hunk 1]
```diff
@@ -77,6 +77,12 @@ hamming_distance [func]
 .. autofunction:: torchmetrics.functional.hamming_distance
     :noindex:
 
+hinge [func]
+~~~~~~~~~~~~~~~~~
```
Comment (Borda) [inline docs/source/references/functional.rst:81]:
```suggestion
~~~~~~~~~~~~
```

File: docs/source/references/modules.rst [hunk 2]
```diff
@@ -162,6 +162,18 @@ FBeta
 .. autoclass:: torchmetrics.FBeta
     :noindex:
 
+HammingDistance
+~~~~~~~~~~~~~~~~
```
Comment (Borda) [inline docs/source/references/modules.rst:166]:
```suggestion
~~~~~~~~~~~~~~~
```

**Comment by ethanwharris** [comment]
@Borda @SkafteNicki Thanks for the comments :smiley: I've added check for `multiclass_mode` and switched hinge loss to just hinge so should all be ready for review now.

File: torchmetrics/classification/hinge.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        dist_sync_fn: Callable = None,
+    ):
+        super().__init__(
+            compute_on_step=compute_on_step,
+            dist_sync_on_step=dist_sync_on_step,
+            process_group=process_group,
+            dist_sync_fn=dist_sync_fn,
+        )
+
+        self.add_state("loss", default=tensor(0.0), dist_reduce_fx="sum")
```
Comment (Borda) [inline torchmetrics/classification/hinge.py:92]:
```suggestion
        self.add_state("measure", default=tensor(0.0), dist_reduce_fx="sum")
```

File: torchmetrics/classification/hinge.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        self.add_state("total", default=tensor(0), dist_reduce_fx="sum")
+
+        if multiclass_mode not in (None, 'crammer_singer', 'one_vs_all'):
```
Comment (Borda) [inline torchmetrics/classification/hinge.py:95]: lets define these as Class static StrEnum

File: torchmetrics/functional/classification/hinge.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    if preds.ndim == 1:
+        if preds.shape != target.shape:
+            raise ValueError(
+                "The `preds` and `target` should have the same shape,",
+                f" got `preds` with shape={preds.shape} and `target` with shape={target.shape}.",
+            )
+        mode = DataType.BINARY
+    elif preds.ndim == 2:
+        if preds.shape[0] != target.shape[0]:
+            raise ValueError(
```
Comment (Borda) [inline torchmetrics/functional/classification/hinge.py:41]: seems we do not have tests for raising invalid values, mind add some

**Review by Borda — COMMENTED** [review]
very nice, almost there, just last comments from me :]

File: torchmetrics/functional/classification/hinge.py (hunk 5 above)
Comment (ethanwharris) [reply torchmetrics/functional/classification/hinge.py:41]: Should be done already here (but codecov is still catching up haha): https://github.com/PyTorchLightning/metrics/blob/cfea3a077801bfb5e4dfa24d2c85a5a7d979d76c/tests/classification/test_hinge.py#L152

File: torchmetrics/classification/hinge.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+            ``MulticlassMode.CRAMMER_SINGER`` or ``"crammer-singer"``, uses the Crammer Singer multi-class hinge loss.
+            ``MulticlassMode.ONE_VS_ALL`` or ``"one-vs-all"`` computes the hinge loss in a one-vs-all fashion.
+
+    Raises:
+        ValueError:
+            If ``multiclass_mode`` is not: None, ``MulticlassMode.CRAMMER_SINGER``, ``"crammer-singer"``,
+            ``MulticlassMode.ONE_VS_ALL`` or ``"one-vs-all"``.
+
+    Example:
+        >>> import torch
```
Comment (SkafteNicki) [inline torchmetrics/classification/hinge.py:64]:
```suggestion
        # binary example
        >>> import torch
```

File: torchmetrics/classification/hinge.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+        >>> from torchmetrics import Hinge
+        >>> target = torch.tensor([0, 1, 1])
+        >>> preds = torch.tensor([-2.2, 2.4, 0.1])
+        >>> hinge = Hinge()
+        >>> hinge(preds, target)
+        tensor(0.3000)
+
+        >>> target = torch.tensor([0, 1, 2])
```
Comment (SkafteNicki) [inline torchmetrics/classification/hinge.py:72]:
```suggestion

        # multiclass example, default mode
        >>> target = torch.tensor([0, 1, 2])
```

File: torchmetrics/classification/hinge.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        >>> preds = torch.tensor([[-1.0, 0.9, 0.2], [0.5, -1.1, 0.8], [2.2, -0.5, 0.3]])
+        >>> hinge = Hinge()
+        >>> hinge(preds, target)
+        tensor(2.9000)
+
+        >>> target = torch.tensor([0, 1, 2])
```
Comment (SkafteNicki) [inline torchmetrics/classification/hinge.py:78]:
```suggestion

        # multiclass example, one vs all mode
        >>> target = torch.tensor([0, 1, 2])
```

File: torchmetrics/functional/classification/hinge.py [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+        ValueError:
+            If preds shape is not of size (N) or (N, C).
+        ValueError:
+            If target shape is not of size (N).
+        ValueError:
+            If ``multiclass_mode`` is not: None, ``MulticlassMode.CRAMMER_SINGER``, ``"crammer-singer"``,
+            ``MulticlassMode.ONE_VS_ALL`` or ``"one-vs-all"``.
+
+    Example:
+        >>> import torch
```
Comment (SkafteNicki) [inline torchmetrics/functional/classification/hinge.py:162]: above comments also applies to these examples

**Review by SkafteNicki — APPROVED** [review]
smaller comments else LGTM

File: torchmetrics/functional/classification/hinge.py (hunk 9 above)
Comment (ethanwharris) [reply torchmetrics/functional/classification/hinge.py:163]: Done :)

**Comment by ethanwharris** [comment]
@PyTorchLightning/core-metrics I was having a weird issue here where the test was hanging forever rather than reporting an error with `ddp` and `dist_sync_on_step` both `True`. Managed to fix the bug so all works now, but ideally it would report the test failure rather than hang. Is this a known issue?

**Review by ananyahjha93 — APPROVED** [review]
